### PR TITLE
IGNITE-19947 Sql. Fix CAST(n AS DECIMAL(precision, scale)) behaviour for scale > precision issues

### PR DESCRIPTION
**Pull request by zstan**
https://issues.apache.org/jira/browse/IGNITE-19947

**SELECT CAST (0.0001100011 AS DECIMAL(3, 6));** correct output : **0.000110**
**SELECT CAST (0.0001110011 AS DECIMAL(2, 6));** not applicable cause after trimming by **scale** : 0.000111 - 3 digits are involved

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java [hunk 1]
```diff
@@ -213,31 +191,38 @@ public static BigDecimal toBigDecimal(Object o, int precision, int scale) {
...
+                : toBigDecimal(o.toString(), precision, scale);
     }
 
-    /**
-     * Converts the given {@code BigDecimal} to a decimal with the given {@code precision} and {@code scale}
-     * according to SQL spec for CAST specification: General Rules, 8.
-     */
-    public static BigDecimal convertDecimal(BigDecimal value, int precision, int scale) {
+    /** Convert and validate input. */
+    public static BigDecimal toBigDecimal(Number value, int precision, int scale) {
```
Comment (lowka) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java:198]: What is wrong with existing comment?

Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java:201]: fixed

**Review by lowka — APPROVED** [review]

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java [hunk 2]
```diff
@@ -371,8 +372,7 @@ private static Stream<Arguments> decimalCastFromLiterals() {
                 // TODO Uncomment these test cases after https://issues.apache.org/jira/browse/IGNITE-19822 is fixed.
                 arguments(CaseStatus.SKIP, numeric, "100.12", decimalType(5, 1), bigDecimalVal("100.1")),
```
Comment (AMashenkov) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java:373]: Let's also validate RoundingMode.
`arguments(CaseStatus.SKIP, numeric, "100.15", decimalType(5, 1), bigDecimalVal("100.2")),`

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java [hunk 3]
```diff
@@ -246,7 +234,86 @@ public static BigDecimal convertDecimal(BigDecimal value, int precision, int sca
...
+            // calculate overall processing digits
+            int digit = scale > processingDigits ? scale - processingDigits : 0;
+            for (int i = strRepr.length() - 1; i >= 0; i--) {
+                digit++;
+                if (strRepr.charAt(i) != '0') {
+                    lastSignificantDigit = digit;
+                }
+            }
+
+            if (lastSignificantDigit > precision) {
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java:292]: Does `lastSignificant` value equals to BigDeciman.ulp().scale() ?

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java (hunk 2 above)
Comment (zstan) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java:373]: done

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java (hunk 3 above)
Comment (zstan) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java:292]: not suitable ... (0.0101) BigDeciman.ulp().scale() == 4, but 3 is avoid here ..

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java [hunk 4]
```diff
@@ -246,7 +234,86 @@ public static BigDecimal convertDecimal(BigDecimal value, int precision, int sca
...
+            // fractional part
+            strRepr = strRepr.substring(pos + 1);
+
+            // length of fractional part
+            int processingDigits = strRepr.length();
+
+            // cut if scale is less than processing digits
+            if (scale < processingDigits) {
+                strRepr = strRepr.substring(0, scale);
+            }
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java:281]: Let's avoid new string creation.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java [hunk 5]
```diff
@@ -213,40 +192,90 @@ public static BigDecimal toBigDecimal(Object o, int precision, int scale) {
...
-            if (scale > precision) {
-                throw new SqlException(RUNTIME_ERR, NUMERIC_FIELD_OVERFLOW_ERROR);
-            } else {
-                int currentSignificantDigits = value.precision() - value.scale();
-                int expectedSignificantDigits = precision - scale;
+        if (dec != null) {
+            return dec;
+        } else {
+            dec = convertToBigDecimal(value);
+        }
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java:225]: Actually, there are 2 cases:
numbers, which contains fractional part only, (when `value.longValue() == 0`)
and numbers with integral part.
```suggestion
        if (value.longValue() == 0) {
            return processFractionalValue(...);
        } else {
            return processValueWithIntegralPart(..);
        }
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java [hunk 6]
```diff
@@ -213,40 +192,90 @@ public static BigDecimal toBigDecimal(Object o, int precision, int scale) {
...
         }
 
-        return value.setScale(scale, RoundingMode.HALF_UP);
+        return dec.setScale(scale, roundingMode);
+    }
+
+    private static @Nullable BigDecimal processFractionData(Number value, int precision, int scale) {
+        if (value.longValue() != 0) {
+            return null;
+        }
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/exp/IgniteSqlFunctions.java:244]: Let's move this check to outside of this method.

**Pull request merged**